evm_arithmetization/src/fixed_recursive_verifier.rs

```diff
@@ -811,16 +830,33 @@ where
             stark_config.num_challenges,
         );
         // Check that the correct CTL challenges are used in every proof.
-        for pi in &pis {
-            for i in 0..stark_config.num_challenges {
-                builder.connect(
-                    ctl_challenges.challenges[i].beta,
-                    pi.ctl_challenges.challenges[i].beta,
-                );
-                builder.connect(
-                    ctl_challenges.challenges[i].gamma,
-                    pi.ctl_challenges.challenges[i].gamma,
-                );
+        for (i, pi) in pis.iter().enumerate() {
+            for j in 0..stark_config.num_challenges {
+                if KECCAK_TABLES_INDICES.contains(&i) {
+                    // Ensures that the correct CTL challenges are used in Keccak tables when
+                    // `enable_keccak_tables` is true.
+                    let beta_diff = builder.sub(
+                        ctl_challenges.challenges[j].beta,
+                        pi.ctl_challenges.challenges[j].beta,
+                    );
+                    let gamma_diff = builder.sub(
+                        ctl_challenges.challenges[j].gamma,
+                        pi.ctl_challenges.challenges[j].gamma,
+                    );
+                    let beta_check = builder.mul(use_keccak_tables.target, beta_diff);
+                    let gamma_check = builder.mul(use_keccak_tables.target, gamma_diff);
+                    builder.assert_zero(beta_check);
+                    builder.assert_zero(gamma_check);
```

Comment:
nit: you can use `conditional_assert_eq`, which behaves like `connect` but conditionally:

```suggestion
                    builder.conditional_assert_eq(
                        use_keccak_tables.target,
                        ctl_challenges.challenges[j].beta,
                        pi.ctl_challenges.challenges[j].beta,
                    );
                    builder.conditional_assert_eq(
                        use_keccak_tables.target,
                        ctl_challenges.challenges[j].gamma,
                        pi.ctl_challenges.challenges[j].gamma,
                    );
```